scripts/ci/pairing/pkg/cmd/pair.go

```diff
@@ -0,0 +1,176 @@
+	opt := &github.PullRequestListOptions{
+		State: "open",
+	}
+
+	prs, _, err := client.PullRequests.List(context.Background(), owner, repo, opt)
+	if err != nil {
+		return nil, fmt.Errorf("could not list pull requests: %w", err)
+	}
+
+	return prs, nil
+}
+
+// shouldPair determines whether the given remote and branch name should be paired
+// based on existing open pull requests in the given organization and repository.
+func shouldPair(orgForPairing, repoForPairing, currentRemoteName, currentBranchName string) (bool, error) {
+	pullRequests, err := listOpenPRs(orgForPairing, repoForPairing)
+	if err != nil {
+		return false, err
+	}
+
+	for _, pull := range pullRequests {
+		if pull.GetHead().GetRef() == currentBranchName && pull.GetUser().GetLogin() == currentRemoteName {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// getCurrentPRInfo gets the current info of the PR that triggered the pairing
+// the pairing can be triggered by CI job or GH action
+func getCurrentPRInfo() (*PullRequestMetadata, error) {
+	pr := &PullRequestMetadata{}
+	jobSpecEnvVarData := os.Getenv("JOB_SPEC")
+
+	// running CI job
+	if jobSpecEnvVarData != "" {
+		log.Println("running in CI job")
+		jobSpec := &JobSpec{}
+		if err := json.Unmarshal([]byte(jobSpecEnvVarData), jobSpec); err != nil {
+			return pr, fmt.Errorf("error when parsing openshift job spec data: %w", err)
+		}
+		if len(jobSpec.Refs.Pulls) == 1 {
+			pull := jobSpec.Refs.Pulls[0]
+			pr = &PullRequestMetadata{
+				RemoteName: pull.Author,
+				Number:     strconv.Itoa(pull.Number),
+				BranchName: pull.HeadRef,
+			}
+		} else {
+			fmt.Println("No pull request data found.")
+		}
+		// running GH action
+	} else if os.Getenv("GITHUB_ACTIONS") != "" {
+		log.Println("running in GH action")
+		pr = &PullRequestMetadata{
+			RemoteName: os.Getenv("AUTHOR"),
+			BranchName: os.Getenv("GITHUB_HEAD_REF"),
+		}
+	}
+
+	return pr, nil
+}
+
+func clone(cloneDir, org, repo, prRemoteName, prBranchName string, p PairingServiceInterface) error {
+	branch := "master"
+
+	cloneDirInfo, err := os.Stat(cloneDir)
+	if !os.IsNotExist(err) && cloneDirInfo.IsDir() {
+		log.Printf("folder %s already exists... removing", cloneDir)
+
+		err := os.RemoveAll(cloneDir)
+		if err != nil {
+			return fmt.Errorf("error removing %s folder: %w", cloneDir, err)
+		}
+	}
+
+	// if CI
+	if prRemoteName != "" && prBranchName != "" {
+		shouldPair, err := p.shouldPair(org, repo, prRemoteName, prBranchName)
+		if err != nil {
+			return err
+		}
+		// check if pairing is required
+		if shouldPair {
+			org = prRemoteName
+			branch = prBranchName
+		}
+	}
+
+	url := fmt.Sprintf("https://github.com/%s/%s", org, repo)
+	refName := fmt.Sprintf("refs/heads/%s", branch)
+
+	log.Printf("cloning '%s' with git ref '%s'", url, refName)
+
+	_, err = git.PlainClone(cloneDir, false, &git.CloneOptions{
+		URL:           url,
+		ReferenceName: plumbing.ReferenceName(refName),
+		Progress:      os.Stdout,
+	})
```

Comment:
When you look at the existing pairing logic, it always clones the upstream repo, not the fork. Then it adds the fork as an external source and fetches the branch. As the last step it merges the master branch into the remove branch to make sure that the paired version of the repo corresponds to the state when the PR is gonna be merged
https://github.com/codeready-toolchain/toolchain-e2e/blob/9bc394e74fcca131d43f1024c5fbd2a6e25781fd/make/ksctl.mk#L53-L59
We need to keep this flow